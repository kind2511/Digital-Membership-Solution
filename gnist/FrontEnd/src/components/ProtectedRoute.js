import React from 'react';
import { useAuth0 } from '@auth0/auth0-react';
import Loading from './Loading';

const ProtectedRoute = ({ component: Component, ...rest }) => {
  const { isAuthenticated, isLoading, loginWithRedirect, error } = useAuth0();

  if (isLoading) {
    return <Loading />;
  }

  if (error) {
    // Handle the error state
    return <div>Oops... {error.message}</div>;
  }

  if (!isAuthenticated) {
    loginWithRedirect(); // Redirect to Auth0 login
    return <Loading />;
  }

  return Component ? <Component {...rest} /> : null; // Render nothing if no component is provided
};

export default ProtectedRoute;
