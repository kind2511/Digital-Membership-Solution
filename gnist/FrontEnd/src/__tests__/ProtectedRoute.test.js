import React from 'react';
import renderer from 'react-test-renderer';
import ProtectedRoute from '../components/ProtectedRoute';
import Loading from '../components/Loading';
import { Auth0Context } from '@auth0/auth0-react';

// Mock the Loading component
jest.mock('../components/Loading', () => () => <div>Loading...</div>);

// Mock a sample component to be protected
const MockComponent = () => <div>Mock Component</div>;

// Helper function to create a tree with given Auth0 context values
const createTree = (auth0ContextValues, props) => {
  return renderer.create(
    <Auth0Context.Provider value={auth0ContextValues}>
      <ProtectedRoute component={MockComponent} {...props} />
    </Auth0Context.Provider>
  ).toJSON();
};

test('ProtectedRoute renders loading state', () => {
  const tree = createTree({ isLoading: true });
  expect(tree).toMatchSnapshot();
});

test('ProtectedRoute handles error state', () => {
  const tree = createTree({ isLoading: false, error: { message: 'Something went wrong' } });
  expect(tree).toMatchSnapshot();
});

test('ProtectedRoute redirects when not authenticated', () => {
  const mockLoginWithRedirect = jest.fn();
  const tree = createTree({ isLoading: false, isAuthenticated: false, loginWithRedirect: mockLoginWithRedirect });
  expect(tree).toMatchSnapshot();
  expect(mockLoginWithRedirect).toHaveBeenCalled();
});

test('ProtectedRoute renders component when authenticated', () => {
  const tree = createTree({ isLoading: false, isAuthenticated: true });
  expect(tree).toMatchSnapshot();
});
