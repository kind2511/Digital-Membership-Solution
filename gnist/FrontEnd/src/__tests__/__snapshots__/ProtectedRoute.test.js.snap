// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute handles error state 1`] = `
<div>
  Oops... 
  Something went wrong
</div>
`;

exports[`ProtectedRoute redirects when not authenticated 1`] = `
<div>
  Loading...
</div>
`;

exports[`ProtectedRoute renders component when authenticated 1`] = `
<div>
  Mock Component
</div>
`;

exports[`ProtectedRoute renders loading state 1`] = `
<div>
  Loading...
</div>
`;
